addons/purchase_stock/models/purchase.py

```diff
@@ -228,22 +230,27 @@ def _get_picking_type(self, company_id):
             picking_type = self.env['stock.picking.type'].search([('code', '=', 'incoming'), ('warehouse_id', '=', False)])
         return picking_type[:1]
 
-    def _prepare_picking(self):
+    def _prepare_picking(self, is_return=False):
         if not self.group_id:
             self.group_id = self.group_id.create({
                 'name': self.name,
                 'partner_id': self.partner_id.id
             })
+        picking_type = self.picking_type_id if not is_return else self.picking_type_id.return_picking_type_id
+        if not picking_type:
+            raise UserError(_("No picking type defined for %s operation.", "return" if is_return else "receipt"))
         if not self.partner_id.property_stock_supplier.id:
             raise UserError(_("You must set a Vendor Location for this partner %s", self.partner_id.name))
+        location_id = self.partner_id.property_stock_supplier.id if not is_return else self._get_destination_location()
+        location_dest_id = self._get_destination_location(is_return)
```

Comment:
We can't rely on `_get_destination_location()` as it doesn't look at the return picking type. To me, it would make more sense to create a new method `_get_locations()` or something similar (ideally in picking.type but I think that's too much for this PR) to return from and to locations, and deprecate `_get_destination_location()` in master if appropriate.